File: src/commands/stats.ts
import {
  type ChatInputCommandInteraction,
  MessageFlags,
  SlashCommandBuilder,
} from "discord.js";
import { and, count, desc, eq, gte, sql } from "drizzle-orm";
import type { Command } from "../client.js";
import { db } from "../db/index.js";
import { inviteUses, trackedInvites } from "../db/schema.js";
import { brandEmbed, errorEmbed } from "../lib/embeds.js";
import { requireGuild } from "../lib/permissions.js";

const PERIOD_MAP: Record<string, number> = {
  "7d": 7,
  "30d": 30,
  all: 0,
};

const command: Command = {
  data: new SlashCommandBuilder()
    .setName("stats")
    .setDescription("View server-wide invite analytics")
    .setDefaultMemberPermissions(0x8)
    .addStringOption((opt) =>
      opt
        .setName("period")
        .setDescription("Time period for stats")
        .addChoices(
          { name: "Last 7 days", value: "7d" },
          { name: "Last 30 days", value: "30d" },
          { name: "All time", value: "all" },
        ),
    ),

  async execute(interaction: ChatInputCommandInteraction) {
    const guildId = await requireGuild(interaction);
    if (!guildId) return;
    const period = interaction.options.getString("period") ?? "30d";
    const days = PERIOD_MAP[period] ?? 30;

    // Total tracked invites
    const [{ value: totalInvites }] = await db
      .select({ value: count() })
      .from(trackedInvites)
      .where(eq(trackedInvites.guildId, guildId));

    if (totalInvites === 0) {
      await interaction.reply({
        embeds: [
          errorEmbed("No tracked invites yet. Use `/create` to get started."),
        ],
        flags: MessageFlags.Ephemeral,
      });
      return;
    }

    // Build date filter for period
    const dateFilter =
      days > 0
        ? gte(
            inviteUses.joinedAt,
            sql`NOW() - INTERVAL '${sql.raw(String(days))} days'`,
          )
        : undefined;

    const baseWhere = and(eq(inviteUses.guildId, guildId), dateFilter);

    // Total joins in period
    const [{ value: totalJoins }] = await db
      .select({ value: count() })
      .from(inviteUses)
      .where(baseWhere);

    const joinsCount = count(inviteUses.id);

    // Top 5 primary sources
    const topPrimary = await db
      .select({
        source: trackedInvites.primarySource,
        joins: joinsCount,
      })
      .from(inviteUses)
      .innerJoin(
        trackedInvites,
        eq(trackedInvites.id, inviteUses.trackedInviteId),
      )
      .where(baseWhere)
      .groupBy(trackedInvites.primarySource)
      .orderBy(desc(joinsCount))
      .limit(5);

    // Top 5 secondary sources
    const topSecondary = await db
      .select({
        source: trackedInvites.secondarySource,
        joins: joinsCount,
      })
      .from(inviteUses)
      .innerJoin(
        trackedInvites,
        eq(trackedInvites.id, inviteUses.trackedInviteId),
      )
      .where(baseWhere)
      .groupBy(trackedInvites.secondarySource)
      .orderBy(desc(joinsCount))
      .limit(5);

    const periodLabel = days > 0 ? `Last ${days} days` : "All time";

    const primaryText =
      topPrimary.length > 0
        ? topPrimary
            .map((r, i) => `${i + 1}. **${r.source}** \u2014 ${r.joins} joins`)
            .join("\n")
        : "No data yet";

    const secondaryText =
      topSecondary.length > 0
        ? topSecondary
            .map((r, i) => `${i + 1}. **${r.source}** \u2014 ${r.joins} joins`)
            .join("\n")
        : "No data yet";

    const embed = brandEmbed()
      .setTitle(`Invite Stats for ${interaction.guild?.name}`)
      .setDescription(`Showing stats for: **${periodLabel}**`)
      .addFields(
        {
          name: "Total Tracked Invites",
          value: String(totalInvites),
          inline: true,
        },
        { name: "Total Joins", value: String(totalJoins), inline: true },
        { name: "\u200b", value: "\u200b", inline: true },
        { name: "Top Primary Sources", value: primaryText, inline: false },
        { name: "Top Secondary Sources", value: secondaryText, inline: false },
      );

    await interaction.reply({ embeds: [embed], flags: MessageFlags.Ephemeral });
  },
};

export default command;

File: src/commands/stats.rs
use anyhow::Context as _;
use chrono::{Duration, Utc};

use crate::commands::support::format_percentage;
use crate::embeds;
use crate::models::SourceCount;
use crate::{Context, Error};

#[derive(Clone, Copy, Debug, poise::ChoiceParameter)]
pub enum StatsPeriod {
    #[name = "Last 7 days"]
    SevenDays,
    #[name = "Last 30 days"]
    ThirtyDays,
    #[name = "All time"]
    AllTime,
}

impl StatsPeriod {
    pub(crate) const fn days(self) -> Option<i64> {
        match self {
            Self::SevenDays => Some(7),
            Self::ThirtyDays => Some(30),
            Self::AllTime => None,
        }
    }

    pub(crate) fn since(self) -> Option<chrono::NaiveDateTime> {
        self.days()
            .map(|days| Utc::now().naive_utc() - Duration::days(days))
    }

    pub(crate) fn label(self) -> String {
        self.days()
            .map_or_else(|| "All time".to_owned(), |days| format!("Last {days} days"))
    }
}

/// View server-wide invite analytics.
#[poise::command(
    slash_command,
    guild_only,
    default_member_permissions = "ADMINISTRATOR"
)]
pub async fn stats(
    ctx: Context<'_>,
    #[description = "Time period for the statistics"] period: Option<StatsPeriod>,
) -> Result<(), Error> {
    let guild_id = ctx.guild_id().context("command is missing a guild ID")?;
    let guild_id_string = guild_id.to_string();
    let total_invites = ctx
        .data()
        .repository
        .count_invites(&guild_id_string)
        .await?;

    let period = period.unwrap_or(StatsPeriod::ThirtyDays);
    let since = period.since();
    let counts = ctx
        .data()
        .repository
        .analytics_counts(&guild_id_string, since)
        .await?;
    let top_primary = ctx
        .data()
        .repository
        .top_primary_sources(&guild_id_string, since, 5)
        .await?;
    let top_secondary = ctx
        .data()
        .repository
        .top_secondary_sources(&guild_id_string, since, None, 5)
        .await?;

    let guild_name = ctx
        .guild()
        .map_or_else(|| "this server".to_owned(), |guild| guild.name.clone());
    let period_label = period.label();
    let attribution_rate = format_percentage(counts.attributed, counts.total);
    let embed = embeds::brand()
        .title(format!("Invite Stats for {guild_name}"))
        .description(format!("Showing statistics for **{period_label}**"))
        .field("Active Tracked Invites", total_invites.to_string(), true)
        .field("Human Joins", counts.total.to_string(), true)
        .field("Attributed", counts.attributed.to_string(), true)
        .field("Unattributed", counts.unattributed.to_string(), true)
        .field("Attribution Rate", attribution_rate, true)
        .field("Automated/System Joins", counts.bots.to_string(), true)
        .field("Top Primary Sources", format_sources(&top_primary), false)
        .field(
            "Top Secondary Sources",
            format_sources(&top_secondary),
            false,
        );
    ctx.send(poise::CreateReply::default().embed(embed).ephemeral(true))
        .await?;
    Ok(())
}

fn format_sources(sources: &[SourceCount]) -> String {
    if sources.is_empty() {
        return "No data yet".to_owned();
    }
    sources
        .iter()
        .enumerate()
        .map(|(index, source)| {
            format!(
                "{}. **{}**: {} joins",
                index + 1,
                source.source,
                source.joins
            )
        })
        .collect::<Vec<_>>()
        .join("\n")
}
